sdk/core/azure-core/inc/azure/core/etag.hpp

```diff
@@ -0,0 +1,194 @@
+        return left.m_value.GetValue().compare(right.m_value.GetValue()) == 0;
+      }
+      else
+      {
+        // Left side is weak ETag
+        if (right.IsWeak())
+        {
+          return left.m_value.GetValue().compare(right.m_value.GetValue()) == 0;
+        }
+        // Trim the weak tag
+        auto leftLength = left.m_value.GetValue().length();
+
+        if (leftLength - 2 != right.m_value.GetValue().length())
+        {
+          return false;
+        }
+        return left.m_value.GetValue().compare(2, leftLength - 2, right.m_value.GetValue()) == 0;
+      }
+      return false;
+    }
+
+  protected:
+    /**
+     * @brief Construct an empty (null) #ETag.
+     */
+    ETag() {}
+
+  public:
+    /**
+     * @brief Construct a #ETag.
+     */
+    explicit ETag(std::string etag) : m_value(std::move(etag)) {}
+
+    /**
+     * @brief Copy assignment.
+     */
+    ETag& operator=(const ETag&) = default;
+
+    /**
+     * @brief Whether @Tag is present.
+     * @return `true` if @ETag has a value, `false` otherwise.
+     */
+    const bool HasValue() { return m_value.HasValue(); }
+
+    /*
+     * @brief Return the contained string value
+     * @return #std::string
+     */
+    const std::string& ToString()
+    {
+      if (!m_value.HasValue())
+      {
+        abort();
+      }
+      return m_value.GetValue();
+    }
+
+    /**
+     * @brief Compare with \p other @ETag for equality.
+     * @param other Other @ETag to compare with.
+     * @return `true` if @ETag instances are equal, `false` otherwise.
+     */
+    bool operator==(const ETag& other) const { return ETagComparison(*this, other); }
+
+    /**
+     * @brief Compare with \p other @ETag for inequality.
+     * @param other Other @ETag to compare with.
+     * @return `true` if @ETag instances are not equal, `false` otherwise.
+     */
+    bool operator!=(const ETag& other) const { return !(*this == other); }
+
+    /**
+     * @brief Compare with \p other @ETag for inequality.
+     * @param other Other @ETag to compare with.
+     * @return `true` if @ETag instances are not equal, `false` otherwise.
+     */
+    bool IsWeak() const
+    {
+      // A null eTag is considered strong
+      if (!m_value)
+      {
+        return false;
+      }
+
+      auto& val = m_value.GetValue();
+      // Shortest valid weak etag is 4
+      //  W/""
+      if (val.length() < 4)
+      {
+        return false;
+      }
+
+      // Valid format is W/""
+      //   Must end with a " in the string
+      if (val[0] == 'W' && val[1] == '/' && val[2] == '"' && val[val.size() - 1] == '"')
+      {
+        return true;
+      }
+
+      return false;
+    }
+
+    /**
+     * @brief @ETag representing everything
+     * @notes The any ETag is *, (unquoted).  It is NOT the same as "*"
+     */
+    static const ETag& Any()
+    {
+      static ETag Any = ETag("*");
+      return Any;
```

Comment:
```suggestion
      static ETag any = ETag("*");
      return any;
```

Reply:
Can we just define a static member? Like

```C++
static ETag Any;
static ETag Null;
```

Reply:
> Can we just define a static member?

We _can_ do so, but users then have to understand the dangers of C++'s hated [static initialization order fiasco](https://en.cppreference.com/w/cpp/language/siof). The order of initialization of variables with static storage duration and dynamic initialization is unspecified, so referring to such variables during dynamic initialization becomes a footgun. This pattern (known as the Meyers' Singleton after being popularized by Scott Meyers) avoids the fiasco via function-local statics which use a different mechanism to ensure they are initialized on the first call to the function.

Reply:
What about other uses of static within the SDK, for example, in extensible enums or Datetime? Do they have a similar issue, and if not, why not?

https://github.com/Azure/azure-sdk-for-cpp/blob/5af134b1032b4df1066ac0da7bb4b79e61e9fb79/sdk/storage/azure-storage-files-datalake/inc/azure/storage/files/datalake/protocol/datalake_rest_client.hpp#L147-L149

https://github.com/Azure/azure-sdk-for-cpp/blob/5af134b1032b4df1066ac0da7bb4b79e61e9fb79/sdk/core/azure-core/inc/azure/core/datetime.hpp#L53

Reply:
Yes, they do. I suppose the simplest way to address this is to forbid using the library before `main` is called and then we can use all of the dynamic initializers we like.

Reply:
Is there an alternative design for extensible enums that doesn't rely on statics? This is how they are defined in the guidelines:
https://azure.github.io/azure-sdk/cpp_implementation.html#enumerations


> I suppose the simplest way to address this is to forbid using the library before `main` is called and then we can use all of the dynamic initializers we like.

How would you document this to let the customers know what not to do? What types of runtime errors or bugs would they see if they did use the library before `main` (if we can share some examples, that will help users recognize the issue)?

Reply:
I know the static initialization order fiasco, but I don't know what to do here. There are many many static members in storage SDK, I really don't want to make all of them a static function. Actually I don't expect anybody to use these static members during program loading. So maybe it's not that big a deal.